### PR TITLE
Fix for client.Repository.Content.GetAllContents for root #800

### PR DESCRIPTION
**Pull request by naveensrinivasan**
The client.Repository.Content.GetAllContents #800  does not allow getting root folders for Branches. This fix would allow the users to pass null/empty strings for the path which will address the issue.

### CODE REVIEW

**Comment by shiftkey** [comment]
While dropping the constraint means that you can get to the root of the repository, I think this behaviour is still unclear. Would it be smarter to have an overload that doesn't take a path?

```csharp
public async Task<IReadOnlyList<RepositoryContent>> GetAllContents(string owner, string name)
```

**Comment by naveensrinivasan** [comment]
So without the path the function would only return the root folder which also takes the `ref` in the argument?

If that is the case then the function would have to be named something like `GetAllContentsOfRoot`
```
public async Task<IReadOnlyList<RepositoryContent>> GetAllContentsOfRoot(string owner, string name,string ref)

```

**Comment by shiftkey** [comment]
@naveensrinivasan you're right, I hadn't thought about the refs side. While we are adding in a lot of overloads here, I think they all end up going through one method - so we should be fine.

**Comment by naveensrinivasan** [comment]
@shiftkey 
``` csharp
public async Task<IReadOnlyList<RepositoryContent>> GetAllContents(string owner, string name, string reference)
```
Cannot have this because there is already another overload with the same parameter types

``` csharp
public async Task<IReadOnlyList<RepositoryContent>> GetAllContents(string owner, string name, string path)
```

What do you want to do?

**Comment by Haacked** [comment]
What if we name these `GetAllContentsByRef` and `GetAllContentsByPath`?

**Comment by naveensrinivasan** [comment]
Sounds good to me!

**Comment by naveensrinivasan** [comment]
Implemented changes. The method names are `GetAllContents` and `GetAllContentsByRef`. I reused the  `GetAllContents` name because in one of the overloads we don't pass the `path` and it does not make sense to have a function name as  `GetAllContentsByPath` and don't have the path in argument.

File: Octokit.Reactive/Clients/IObservableRepositoryContentsClient.cs [hunk 1]
```diff
@@ -124,7 +134,19 @@ public interface IObservableRepositoryContentsClient
         /// <returns>
         /// A collection of <see cref="RepositoryContent"/> representing the content at the specified path
         /// </returns>
-        IObservable<RepositoryContent> GetAllContents(string owner, string name, string path, string reference);
+        IObservable<RepositoryContent> GetAllContentsByRef(string owner, string name, string path, string reference);
```
Comment (Haacked) [inline Octokit.Reactive/Clients/IObservableRepositoryContentsClient.cs]: I would reorder `path` and `reference`. The reason is the other overload to this method takes in `owner`, `name`, `reference`. If the user is calling that overload, then decides, no, they want to call this one, they only need to tack on the `path` parameter. They don't need to re-order parameters. Based on my experience, this is a common source of API call errors.

File: Octokit.Reactive/Clients/ObservableRepositoryContentsClient.cs [hunk 2]
```diff
@@ -166,6 +166,25 @@ public IObservable<RepositoryContent> GetAllContents(string owner, string name,
...
+        /// </returns>
+        public IObservable<RepositoryContent> GetAllContents(string owner, string name)
+        {
+            Ensure.ArgumentNotNullOrEmptyString(owner, "owner");
+            Ensure.ArgumentNotNullOrEmptyString(name, "name");
+
+            return _client
+                .Connection
+                .GetAndFlattenAllPages<RepositoryContent>(ApiUrls.RepositoryContent(owner, name, string.Empty));
+
```
Comment (Haacked) [inline Octokit.Reactive/Clients/ObservableRepositoryContentsClient.cs]: Remove extraneous newline here.

**Comment by Haacked** [comment]
:bread: Back to you.

**Comment by naveensrinivasan** [comment]
Done with the changes.

**Comment by Haacked** [comment]
Hmm, @naveensrinivasan, the branch still has conflicts that need to be resolved.

**Comment by shiftkey** [comment]
Looks like some extra changes are in here. I've extracted the relevant commits to #1064 to bring this up to date.

File: Octokit.Reactive/Clients/IObservableRepositoryContentsClient.cs (hunk 1 above)
Comment (tyadav492) [reply Octokit.Reactive/Clients/IObservableRepositoryContentsClient.cs]: path parameter is not accepting any special char like '#' and space. I am trying to retrieve files form directory and i am getting not found exception and under same root directory it is getting all file which does not have any special char.
I have two sub directories : 1) C# Code  and 2) TestApp. i am able to access TestApp but not "C# Code" as it is throwing not found exception. suggest me any solution for this as i am not able to rename the directory sue to some reason.

**Pull request closed without merging**